### PR TITLE
Update go version to 1.21

### PR DESCRIPTION
**Pull request by mabhi**
## Change Overview

This PR switches to using golang 1.21-bullseye (stable).

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [ ] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues 

- fixes #issue-number

## Test Plan
Build image logs: https://github.com/kanisterio/kanister/actions/runs/6491734714/job/17629476802
```
Run docker/build-push-action@0565240e2d4ab88bba5387d719585280857ece09
GitHub Actions runtime token ACs
Docker info
Proxy configuration
Buildx version
....
....
Metadata
  {
    "containerimage.descriptor": {
      "mediaType": "application/vnd.oci.image.index.v1+json",
      "digest": "sha256:31e73cccd7c9debdae7f7c127aaf57694f3780fe087fe2119911e52850fcb3f6",
      "size": 856
    },
    "containerimage.digest": "sha256:31e73cccd7c9debdae7f7c127aaf57694f3780fe087fe2119911e52850fcb3f6",
    "image.name": "ghcr.io/kanisterio/build:bump-go-version-1-21,ghcr.io/kanisterio/build:2023.10.12-0540,ghcr.io/kanisterio/build:v0.0.25,ghcr.io/kanisterio/build:sha-452d015"
  }
```

```
abhijeet@abhijit-mukherjee ~/Documents/mydev/veeam/kanister (bump-go-version-1-21)$ make shell
launching a shell in the containerized build environment
Running build container in interactive shell mode...
root@abhijit-mukherjee:/go/src/github.com/kanisterio/kanister# go version
go version go1.21.3 linux/amd64
root@abhijit-mukherjee:/go/src/github.com/kanisterio/kanister# golangci-lint version
golangci-lint has version v1.51.2 built from 3e8facb4 on 2023-02-19T22:04:30Z
```

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: docker/build/Dockerfile [hunk 1]
```diff
@@ -17,7 +16,7 @@ COPY --from=goreleaser/goreleaser:v1.12.3 /usr/bin/goreleaser /usr/local/bin/
 
 COPY --from=alpine/helm:3.12.2 /usr/bin/helm /usr/local/bin/
 
-COPY --from=golangci/golangci-lint:v1.50 /usr/bin/golangci-lint /usr/local/bin/
+COPY --from=golangci/golangci-lint:v1.51.2 /usr/bin/golangci-lint /usr/local/bin/
```
Comment (PrasadG193) [inline docker/build/Dockerfile:19]: Why do we need to upgrade `golangci-lint`? If we need to upgrade, we should switch to the latest version [1.54.2](https://github.com/golangci/golangci-lint/releases/tag/v1.54.2)

**Review by PrasadG193 — COMMENTED** [review]
Please add build image logs and the test plan.
Thanks

File: docker/build/Dockerfile (hunk 1 above)
Comment (mabhi) [reply docker/build/Dockerfile:19]: I tried with 1.54.2 but it was reporting false positives due to the `unparam` linter hence had to go back.
```
pkg/controllers/repositoryserver/handler.go:344:98: (*RepoServerHandler).setupKopiaRepositoryServer - result 0 (sigs.k8s.io/controller-runtime/pkg/reconcile.Result) is always nil (unparam)
func (h *RepoServerHandler) setupKopiaRepositoryServer(ctx context.Context, logger logr.Logger) (ctrl.Result, error) {

```

**Comment by pavannd1** [comment]
We might have to run `make codegen` once

**Comment by mabhi** [comment]
> We might have to run `make codegen` once

Ran codegen as suggested. No changes reported apart from comments about replacing `Copyright 2023 The Kanister Authors.` to `Copyright 2023 The Kubernetes Authors` in the autogenerated files. Hence ignored.

File: docker/build/Dockerfile (hunk 1 above)
Comment (PrasadG193) [reply docker/build/Dockerfile:19]: Sounds good! Could you please create a github issue and add a link here. We can resolve the conversation after that

**Review by pavannd1 — APPROVED** [review]
LGTM.
Let's create an issue for linter upgrade and merge this.

File: docker/build/Dockerfile (hunk 1 above)
Comment (mabhi) [reply docker/build/Dockerfile:19]: https://github.com/kanisterio/kanister/issues/2408

**Review by PrasadG193 — APPROVED** [review]

**Pull request merged**